File: pixai_proxy.js
/**
 * PixAI API 代理服务器 (Node.js 版本)
 * 用于绕过浏览器 CORS 限制
 */

const http = require('http');
const https = require('https');
const url = require('url');

const PORT = 5555;
const PIXAI_API_BASE = 'https://api.pixai.art';

/**
 * 解析请求体
 */
function parseBody(req) {
  return new Promise((resolve, reject) => {
    let body = '';
    req.on('data', chunk => {
      body += chunk.toString();
    });
    req.on('end', () => {
      try {
        resolve(body ? JSON.parse(body) : {});
      } catch (e) {
        resolve({});
      }
    });
    req.on('error', reject);
  });
}

/**
 * 转发请求到 PixAI API
 */
function proxyRequest(path, method, headers, body, res) {
  // The client sends the token in 'x-api-key'. The API expects 'Authorization: Bearer <token>'.
  // We'll check both 'x-api-key' and 'authorization' for robustness.
  const apiToken = headers['x-api-key'] || headers['authorization'];

  const options = {
    hostname: 'api.pixai.art',
    port: 443,
    path: path,
    method: method,
    headers: {
      'Content-Type': 'application/json',
      // Format the header as "Bearer <token>" if the token exists.
      // Also, remove "Bearer " from the token if it's already there to avoid "Bearer Bearer ...".
      'Authorization': apiToken ? `Bearer ${apiToken.replace(/^Bearer\s+/, '')}` : '',
      'User-Agent': 'SillyTavern-PixAI-Extension/1.0',
    },
  };

  const proxyReq = https.request(options, proxyRes => {
    let data = '';

    proxyRes.on('data', chunk => {
      data += chunk;
    });

    proxyRes.on('end', () => {
      // 设置 CORS 头
      res.writeHead(proxyRes.statusCode, {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type, Authorization',
      });
      res.end(data);
    });
  });

  proxyReq.on('error', error => {
    console.error('代理请求错误:', error);
    res.writeHead(500, {
      'Content-Type': 'application/json',
      'Access-Control-Allow-Origin': '*',
    });
    res.end(JSON.stringify({ error: error.message }));
  });

  if (body && Object.keys(body).length > 0) {
    proxyReq.write(JSON.stringify(body));
  }
  proxyReq.end();
}

/**
 * 创建 HTTP 服务器
 */
const server = http.createServer(async (req, res) => {
  const parsedUrl = url.parse(req.url, true);
  const path = parsedUrl.pathname;
  const method = req.method;

  // 处理 OPTIONS 预检请求
  if (method === 'OPTIONS') {
    res.writeHead(200, {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
      'Access-Control-Allow-Headers': 'Content-Type, Authorization, x-api-key',
    });
    res.end();
    return;
  }

  // 健康检查端点
  if (path === '/health') {
    res.writeHead(200, {
      'Content-Type': 'application/json',
      'Access-Control-Allow-Origin': '*',
    });
    res.end(JSON.stringify({ status: 'ok', message: 'PixAI Proxy is running' }));
    return;
  }

  // 转发到 PixAI API - 支持 /pixai/* 和 /v1/* 路径
  let apiPath = path;

  // 将 /pixai/* 转换为 /v1/*
  if (path.startsWith('/pixai/')) {
    apiPath = path.replace('/pixai/', '/v1/');
  }

  if (path.startsWith('/v1/') || path.startsWith('/pixai/')) {
    try {
      const body = await parseBody(req);
      console.log(`[${new Date().toISOString()}] ${method} ${path} -> ${apiPath}`);
      proxyRequest(apiPath, method, req.headers, body, res);
    } catch (error) {
      console.error('解析请求体错误:', error);
      res.writeHead(400, {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
      });
      res.end(JSON.stringify({ error: 'Invalid request body' }));
    }
    return;
  }

  // 404
  res.writeHead(404, {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
  });
  res.end(JSON.stringify({ error: 'Not found' }));
});

// 启动服务器
server.listen(PORT, '127.0.0.1', () => {
  console.log('========================================');
  console.log('  PixAI API 代理服务器已启动');
  console.log('========================================');
  console.log(`  监听地址: http://127.0.0.1:${PORT}`);
  console.log(`  健康检查: http://127.0.0.1:${PORT}/health`);
  console.log('========================================');
  console.log('  按 Ctrl+C 停止服务器');
  console.log('========================================\n');
});

// 优雅关闭
process.on('SIGINT', () => {
  console.log('\n正在关闭服务器...');
  server.close(() => {
    console.log('服务器已关闭');
    process.exit(0);
  });
});

process.on('SIGTERM', () => {
  console.log('\n正在关闭服务器...');
  server.close(() => {
    console.log('服务器已关闭');
    process.exit(0);
  });
});

File: pixai_proxy.py
#!/usr/bin/env python3
"""
PixAI API 代理服务器
用于解决浏览器 CORS 限制问题
"""

from flask import Flask, request, jsonify
from flask_cors import CORS
import requests
import sys

app = Flask(__name__)
CORS(app)  # 允许所有来源的跨域请求

PIXAI_API_BASE = "https://api.pixai.art/v1"

@app.route('/pixai/task', methods=['POST', 'OPTIONS'])
def create_task():
    """创建 PixAI 任务"""
    if request.method == 'OPTIONS':
        return '', 204
    
    try:
        # 从请求中获取数据
        data = request.get_json()
        api_key = request.headers.get('X-API-Key')
        
        if not api_key:
            return jsonify({'error': 'Missing API key'}), 401
        
        # 转发请求到 PixAI API
        headers = {
            'Authorization': f'Bearer {api_key}',
            'Content-Type': 'application/json'
        }
        
        response = requests.post(
            f'{PIXAI_API_BASE}/task',
            headers=headers,
            json=data,
            timeout=30
        )
        
        return jsonify(response.json()), response.status_code
        
    except Exception as e:
        print(f"Error in create_task: {e}", file=sys.stderr)
        return jsonify({'error': str(e)}), 500


@app.route('/pixai/task/<task_id>', methods=['GET', 'OPTIONS'])
def get_task(task_id):
    """获取任务状态"""
    if request.method == 'OPTIONS':
        return '', 204
    
    try:
        api_key = request.headers.get('X-API-Key')
        
        if not api_key:
            return jsonify({'error': 'Missing API key'}), 401
        
        headers = {
            'Authorization': f'Bearer {api_key}'
        }
        
        response = requests.get(
            f'{PIXAI_API_BASE}/task/{task_id}',
            headers=headers,
            timeout=30
        )
        
        return jsonify(response.json()), response.status_code
        
    except Exception as e:
        print(f"Error in get_task: {e}", file=sys.stderr)
        return jsonify({'error': str(e)}), 500


@app.route('/pixai/media/<path:media_url>', methods=['GET', 'OPTIONS'])
def get_media(media_url):
    """获取图片（转换为 base64）"""
    if request.method == 'OPTIONS':
        return '', 204
    
    try:
        # 下载图片
        response = requests.get(media_url, timeout=30)
        
        if response.status_code == 200:
            import base64
            base64_data = base64.b64encode(response.content).decode('utf-8')
            return jsonify({
                'data': base64_data,
                'format': 'png'
            })
        else:
            return jsonify({'error': 'Failed to download image'}), response.status_code
            
    except Exception as e:
        print(f"Error in get_media: {e}", file=sys.stderr)
        return jsonify({'error': str(e)}), 500


if __name__ == '__main__':
    print("=" * 60)
    print("🚀 PixAI 代理服务器启动中...")
    print("=" * 60)
    print("监听地址: http://127.0.0.1:5555")
    print("按 Ctrl+C 停止服务器")
    print("=" * 60)
    
    app.run(host='127.0.0.1', port=5555, debug=False)

